kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/SignalManager.java

```diff
@@ -0,0 +1,229 @@
+ */
+public class SignalManager {
+
+  private final Path signalDirectory;
+  private final Configuration conf;
+
+  private static final String UNBOUNDED_CONSTRAINT = "unbounded";
+
+  private static final String STAGED_SIGNAL_POSTFIX = ".tmp";
+
+  /**
+   * Creates a new signal manager using the given signal directory.
+   *
+   * @param conf the Hadoop configuration
+   * @param signalDirectory directory in which the manager
+   *                        stores signals.
+   *
+   * @return a signal manager instance.
+   */
+  public static SignalManager create(Configuration conf, Path signalDirectory) {
+    try {
+      FileSystem rootFileSystem = signalDirectory.getFileSystem(conf);
+
+      rootFileSystem.mkdirs(signalDirectory);
+
+      return new SignalManager(signalDirectory, conf);
+    } catch (IOException e) {
+      throw new DatasetIOException("Unable to create signal manager directory: "
+              + signalDirectory, e);
+    }
+  }
+
+  /**
+   * Loads a signal manager that stores metadata data under the given signal
+   * metadata directory if it exists. Returns <code>null</code> if it does not.
+   *
+   * @param conf the Hadoop configuration
+   * @param signalDirectory directory in which the manager stores signals.
+   *
+   * @return a signal manager instance, or <code>null</code> if the given
+   * directory does not exist.
+   */
+  public static SignalManager load(Configuration conf, Path signalDirectory) {
+    try {
+      FileSystem rootFileSystem = signalDirectory.getFileSystem(conf);
+
+      if (rootFileSystem.exists(signalDirectory)) {
+        return new SignalManager(signalDirectory, conf);
+      } else {
+        return null;
+      }
+    } catch (IOException e) {
+      throw new DatasetIOException ("Cannot load signal manager at:"
+              + signalDirectory, e);
+    }
+  }
+
+  private SignalManager(Path signalDirectory, Configuration conf)
+          throws IOException  {
+    this.signalDirectory = signalDirectory;
+    this.conf = conf;
+  }
+
+  /**
+   * Create a signal for the specified constraints.
+   *
+   * @param viewConstraints The constraints to create a signal for.
+   *
+   * @throws DatasetException if the signal could not be created.
+   */
+  public void signalReady(Constraints viewConstraints) {
+    String normalizedConstraints = getNormalizedConstraints(viewConstraints);
+
+    Path signalPath = new Path(signalDirectory, normalizedConstraints);
+    Path stagedSignalPath = new Path(signalDirectory, normalizedConstraints + STAGED_SIGNAL_POSTFIX);
+
+    try{
+      // create the output stream to overwrite the current contents, if the directory or file
+      // does not exist it will be created
+      FileContext context = FileContext.getFileContext(conf);
+      final FSDataOutputStream os = context.create(stagedSignalPath,
+              EnumSet.of(CreateFlag.CREATE, CreateFlag.OVERWRITE), CreateOpts.perms(FsPermission.getDefault()));
```

Comment:
Looks like we might need to go back to using the filesystem rather than the file context, the travis.ci build failed on the hadoop1 profile with CreateFlag not being available